lib/Conversion/ATenToXTenPass.cpp

```diff
@@ -62,58 +64,96 @@ SmallVector<Operation*> getPreviousOps(mlir::Operation* op) {
   return ancestors;
 }
 
-// Adapter fun because tablegen can only bind result and not the op itself.
-bool fuseFirstC2dInTensorAdd(OpResult left, OpResult right) {
+// Return true if op belongs to neural network
+bool isNeuralNetworkOp(Operation *op) {  
+  return mlir::isa<xilinx::xten::Conv2dOp>(op) ||
+      mlir::isa<xilinx::xten::Conv2dReLUOp>(op) ||
+      mlir::isa<xilinx::xten::Conv2dLReLUOp>(op);
+}
+
+// Tablegen adapter to check which op to fuse
+//
+// Traverse graph backwards and return true if the left op has a longer
+// branch than the right op. The length of the op branches is determined
+// by the first common ancestor.
+bool getLongestBranch(OpResult left, OpResult right) {
 
   // Create map to find common ancestor node
   SmallVector<Operation *> worklist;
-  DenseMap<Operation *, unsigned> worklistMap;
+  DenseMap<Operation *, unsigned> opDistanceMap;
 
   // Get left and right operation
   Operation *leftOp = left.getOwner();
   Operation *rightOp = right.getOwner();
 
+  // Ignore case where both ops are the same
+  if (leftOp == rightOp)
+    return true;
+
   // Initialize starting operation based on left op
   worklist.push_back(leftOp);
-  worklistMap.insert(std::pair<Operation*, unsigned>(leftOp, 0));
+  opDistanceMap.insert(std::pair<Operation *, unsigned>(leftOp, 0));
+
+  // TODO: Rewrite using lambda expression
 
   // Iterate backwards over left op branch
   while (!worklist.empty()) {
-    // Get next worklist element
-    Operation* nextOp = *(worklist.begin());
-    unsigned nextLen = worklistMap[nextOp];
-
-    // Get ancestor ops and push to worklist
-    SmallVector<Operation*> prevOps = getPreviousOps(nextOp);
-    for (Operation* op : prevOps) {
-      worklist.push_back(op);
-      worklistMap.insert(std::pair<Operation*, unsigned>(op, nextLen + 1));
+    auto curOp = opDistanceMap.find(worklist.front());
+    SmallVector<Operation *> inputOps = getInputOps(curOp->first);
+    for (Operation *inputOp : inputOps) {
+
+      // TODO: Refactor classification at a later stage
+
+      // // Ops without operands can be ignored for further processing
+      // if (inputOp->getOperands().size() == 0)
+      //   continue;
+
+      // // Ignore list construct op
+      // if (mlir::isa<Torch::PrimListConstructOp>(inputOp))
+      //   continue;
+
+      // Only increment distance if neural network op
+      if (isNeuralNetworkOp(inputOp)) {
```

Comment:
This strikes me as fragile: any operations along a branch that are not classified as `isNeuralNetworkOp` will break the chain as their inputs will not be added to the `worklist`. Can you add some more cases to `test/` that illustrate what you wish to handle here?

Reply:
Update using exclusion list and added test cases.